### PR TITLE
Setup async-local context that lasts the request lifecycle

### PR DESCRIPTION
**Pull request by devinivy**
It's a hack alright! ¯\\\_(ツ)\_/¯

### CODE REVIEW

File: async-local-logger.js [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+        }
+
+        srv.on('request', (req, res) => asyncLocal.run(req, listener, req, res));
+        srv.off('request', listener);
+    })();
+
+    return srv;
+};
+
+internals.storeMap = new WeakMap();
```
Comment (wswoodruff) [inline async-local-logger.js:63]: When the request is sent thru the server and internal hapi code doesn't reference it anywhere anymore, `WeakMap` will allow the key/val pairs for that request to be picked up by garbage collection whenever! Nice!! I had to read up on this when I saw it loLL good stuff 🍭

File: async-local-logger.js [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+        let listener;
+
+        while (!listener) {
+
+            const [event, maybeListener] = await Toys.event(srv, 'newListener', {
+                error: false,
+                multiple: true
+            });
+
+            if (event === 'request') {
```
Comment (wswoodruff) [inline async-local-logger.js:51]: So does this while loop spin until the first request gets in? And from then on it swaps the listener on each request using those srv `request` listeners?

Comment (wswoodruff) [reply async-local-logger.js:51]: Oh it doesn't _really_ _spin_ cuz it's waiting for `newListener` right there gotcha

Comment (devinivy) [reply async-local-logger.js:51]: It waits until hapi registers its `'request'` listener, then swaps that listener out for the same listener, but wrapped in an async-local context.

Comment (devinivy) [reply async-local-logger.js:51]: (And that happens exactly once, not per request.)

Comment (wswoodruff) [reply async-local-logger.js:51]: Oh gotcha! I was assuming `srv.on('request'` acted like our hapi lifecycle hook

Comment (wswoodruff) [reply async-local-logger.js:51]: Oh wait does it? I'm confused — is `asyncLocal.run` down below being run for every request?

**Review by wswoodruff — APPROVED** [review]
Nuts! It works!

🎢👾

**Pull request merged**